cogs/images.py

```diff
@@ -20,45 +20,80 @@
 from discord.ext import commands
 
 # misc imports
+import concurrent.futures
 import os
 import numpy as np
 import cv2
 import math
-from functools import wraps
+from functools import wraps, partial
 from io import BytesIO
 
+MAX_IMAGE_SIZE = 8 * (10**6)
+
 
 def filter_image(func):
     @wraps(func)
     async def wrapper(self, ctx, *args):
         att = await Images.get_attachment(ctx)
         if att is None:
-            await ctx.send(
-                "no image could be found (only attached image files"
-                " can be detected) or message could not be found",
-                delete_after=15)
+            await ctx.send("no image could be found (only attached image files"
+                           " can be detected) or message could not be found")
             return
 
         await ctx.trigger_typing()
-
         buffer = BytesIO()
         await att.save(fp=buffer)
+
         original_name, ext = att.filename.rsplit('.', 1)
+        ext = ext.lower()
+        if ext in ("jpeg", "jpg"):
+            is_png = False
+        elif ext in ("png"):
+            is_png = True
+        else:
+            await ctx.send("image format not supported.")
+            return
+
         fn = f"{original_name}-{func.__name__}.{ext}"
+        ratio = (MAX_IMAGE_SIZE /
+                 att.size) * 100 if att.size > MAX_IMAGE_SIZE else 100
 
         try:
-            img_bytes = np.asarray(bytearray(buffer.read()), dtype=np.uint8)
-            result = cv2.imdecode(img_bytes, cv2.IMREAD_UNCHANGED)
-            args = (args[0], result) if len(args) == 2 else (result, )
-            result = await func(self, ctx, *args)
-            _r, buffer = cv2.imencode(f".{ext}", result,
-                                      [cv2.IMWRITE_JPEG_QUALITY, 100])
+
+            with concurrent.futures.ProcessPoolExecutor() as pool:
```

Comment:
I'd prefer thread pool instead of process pool, plus you should initialize this resource at bot startup (assign a bot.cpu_thread_pool or something). We should not depend on the function to implement that

Reply:
python documentation states that `ProcessPoolExecutor`s should be used for CPU bound tasks (see: https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor)
i don't think that having a `ProcessPoolExecutor` running the entire time is worth it either, considering the overhead, because the image functions aren't used that often, whereas the cost of starting a pool isnt that high anyways, especially considering the opencv2 functions are what take up the majority of the time/resources in these commands

Reply:
after doing some more testing, it seems like the overhead on keeping a `ProcessPoolExecutor` in the background really is minimal, so ill change that